### ISSUE TITLE
trial can't load two test methods from different test cases which inherit the method from a shared base class

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3383]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3383|
|Type|defect|
|Created|2008-08-04 19:08:14Z|
|Branch|https://github.com/twisted/twisted/tree/uniquer-loader-3383|

Consider:

```
exarkun@charm:/tmp$ cat > test_foo.py
from twisted.trial.unittest import TestCase

class X:
        def test_foo(self):
                pass

class A(TestCase, X):
        pass

class B(TestCase, X):
        pass
exarkun@charm:/tmp$ trial test_foo.py
test_foo
  A
    test_foo ...                                                           [OK]
  B
    test_foo ...                                                           [OK]

-------------------------------------------------------------------------------
Ran 2 tests in 0.002s

PASSED (successes=2)
exarkun@charm:/tmp$ trial test_foo.A.test_foo test_foo.B.test_foo
test_foo
  B
    test_foo ...                                                           [OK]

-------------------------------------------------------------------------------
Ran 1 tests in 0.002s

PASSED (successes=1)
exarkun@charm:/tmp$
```

This is because `loadByNames` puts candidates into a `set` and the two unbound methods named above hash and compare equal to each other.

Attachments:

* [3383.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4a7/4a755b4c5b406166da2a770c53b49e61f03e1cce/9ea825419fe359bf9dc9fda2fe779ebb09b986ec.patch) (2806 bytes) - added by djfroofy on 2011-03-15 16:51:20Z - 
* [3383-r1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4a7/4a755b4c5b406166da2a770c53b49e61f03e1cce/376631c33ec26c8240d86e39ca156a846dd1a656.patch) (3335 bytes) - added by djfroofy on 2011-03-16 16:43:32Z - 
* [3383-r2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4a7/4a755b4c5b406166da2a770c53b49e61f03e1cce/e19063de5c5efd4254c18df04ec80a965adf5ec9.patch) (3338 bytes) - added by djfroofy on 2011-03-21 19:10:35Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3383 3383
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_uniquer_loader_3383 branches/uniquer-loader-3383
branch_author__djfroofy djfroofy
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1217876894000000 1217876894000000
changetime__1300755930000000 1300755930000000
version__None None
owner__exarkun exarkun
cc__djfroofy
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)|@djfroofy set owner to @djfroofy<br>@djfroofy set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)<a name="note_3"></a>|@djfroofy removed owner<br>@djfroofy set status to `new`|
|-|-|

The patch uses custom method `_uniqueTest()` instead of `set()` to gather unique list of tests.  This treats methods specially and makes a stand-in hash over the 2-tuple `(method.im_class, method)`.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_4"></a>|@washort set owner to @djfroofy|
|-|-|

The test in this patch passes without the implementation change. There should be a test that fails in order to validate the bugfix.

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)|@djfroofy set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)<a name="note_6"></a>|@djfroofy removed owner<br>@djfroofy set status to `new`|
|-|-|

Yes, sorry. I see now the previous test was very wrong.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml commented|
|-|-|

How does this interact with this chunk of code in t/trial/unittest.py? 

```
    if sys.version_info >= (2, 6):
        # Override the comparison defined by the base TestCase which considers
        # instances of the same class with the same _testMethodName to be
        # equal.  Since trial puts TestCase instances into a set, that
        # definition of comparison makes it impossible to run the same test
        # method twice.  Most likely, trial should stop using a set to hold
        # tests, but until it does, this is necessary on Python 2.6.  Only
        # __eq__ and __ne__ are required here, not __hash__, since the
        # inherited __hash__ is compatible with these equality semantics.  A
        # different __hash__ might be slightly more efficient (by reducing
        # collisions), but who cares? -exarkun
        def __eq__(self, other):
            return self is other

        def __ne__(self, other):
            return self is not other
```

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)<a name="note_8"></a>|@djfroofy commented|
|-|-|

I don't think there is any new interaction here. Nothing is dynamically changed on the objects which are passed to the `TestSuite`, they are just collected differently for this specific case: `loadByNames()`.

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_9"></a>|@moshez set owner to @djfroofy|
|-|-|

Two comments on the code:
* Why is it putting the thing second in the tuple, and then using [-1]? Wouldn't it be just as nice to use [0] and putting the thing first? It would look nicer that way
* Why are you not using isinstance? I mean, yes, isinstance is in general to be avoided, but in this case, type() is just as bad (and this would work well with weird-ass metaclasses which overload MethodType -- and there's no downside)

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)<a name="note_10"></a>|@djfroofy commented|
|-|-|

Replying to [moshez](#note_9):
> * Why is it putting the thing second in the tuple, and then using [-1]? Wouldn't it be just as nice to use [0] and putting the thing first? It would look nicer that way

No particular reason, but I've changed with latest patch to use [0].

> * Why are you not using isinstance? I mean, yes, isinstance is in general to be avoided, but in this case, type() is just as bad (and this would work well with weird-ass metaclasses which overload MethodType -- and there's no downside)

I wasn't thinking about such weird-ass metaclasses, but I guess `isinstance()` should cover this. I also agree that in general it's bad to use `isinstance` or similar type checking - but what's the better thing to do for this case?

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)|@djfroofy removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [31374]) Branching to 'uniquer-loader-3383'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [31375]) Apply 3383-r2.patch 

refs [#3383](https://github.com/twisted/twisted/issues/3383)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

Cool.  I tweaked some whitespace a little (r31376) and added a news fragment (r31377).  Aside from those things, the latest patch looks great, and the [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/uniquer-loader-3383) look good.  I'll merge this.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `closed`|
|-|-|

(In [31378]) Merge uniquer-loader-3383

Author: djfroofy
Reviewer: washort, moshez, exarkun
Fixes: [#3383](https://github.com/twisted/twisted/issues/3383)


Be careful with methods when loading tests to run in trial.  Ensure that
if one test method is inherited by more than one class, that test method
is run for each class that inherits it.
